### PR TITLE
feat: Added api endpoint for hotel (BREAD)

### PR DESCRIPTION
**Pull request by Anders164a**
## Description

- Added api endpoint for hotel (BREAD)
- Added tests for hotel (BREAD)

## Type of Change

- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: src/modules/hotel/__test__/createHotel.test.ts [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+            error: "Bad Request",
+            message: "body must have required property 'name'",
+            statusCode: 400,
+        });
+    });
+
+    it("should return status 400, when name is empty", async () => {
+        const response = await fastify.inject({
+            method: "POST",
+            url: "/api/hotel/",
```
Comment (kristianbinau) [inline src/modules/hotel/__test__/createHotel.test.ts:117]: 🤮 Remove trailing '/' everywhere. There's atleast 13 places where this is happening.

File: src/modules/hotel/hotel.controller.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        return reply.badRequest(await error_message(e));
+    }
+}
+
+export async function browseHotelHandler(
+    request: FastifyRequest,
+    reply: FastifyReply
+) {
+    try {
+        const hotels = await browseHotel();
```
Comment (kristianbinau) [inline src/modules/hotel/hotel.controller.ts:37]: Should be cached with redis

File: src/modules/hotel/hotel.service.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+    return new_hotel;
+}
+
+export async function deleteHotel(id: number) {
+    const hotel = await findHotelById(id);
+    if (!hotel) {
+        return id_not_found(id);
+    }
+
```
Comment (kristianbinau) [inline src/modules/hotel/hotel.service.ts:58]: No need to lookup hotel before trying to delete. If the hotel is not found then its already deleted.

File: src/modules/hotel/hotel.controller.ts (hunk 2 above)
Comment (kristianbinau) [reply src/modules/hotel/hotel.controller.ts:37]: All gets should be cached, and invalidated on update and delete

**Review by kristianbinau — CHANGES_REQUESTED** [review]
First batch

File: prisma/schema.prisma [hunk 4]
```diff
@@ -20,3 +20,12 @@ model User {
     createdAt DateTime @default(now())
     updatedAt DateTime @updatedAt
 }
+
+model Hotel {
+    id          Int      @id @default(autoincrement())
+    name        String   @db.VarChar(50)
```
[the remark this answers is not in the document]
Comment (Anders164a) [reply prisma/schema.prisma:26]: Changes added

File: src/modules/hotel/__test__/createHotel.test.ts (hunk 1 above)
Comment (Anders164a) [reply src/modules/hotel/__test__/createHotel.test.ts:117]: Changes added

File: src/modules/hotel/hotel.service.ts (hunk 3 above)
Comment (Anders164a) [reply src/modules/hotel/hotel.service.ts:58]: Changes added

File: src/modules/hotel/hotel.controller.ts (hunk 2 above)
Comment (Anders164a) [reply src/modules/hotel/hotel.controller.ts]: Changes added

**Review by kristianbinau — CHANGES_REQUESTED** [review]
Some more

File: src/modules/hotel/hotel.controller.ts [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+export async function createHotelHandler(
+    request: FastifyRequest<{
+        Body: CreateHotelInput;
+    }>,
+    reply: FastifyReply
+) {
+    try {
+        const hotel = await createHotel(request.body);
+
```
[the remark this answers is not in the document]
Comment (Anders164a) [reply src/modules/hotel/hotel.controller.ts:26]: Changes added

File: src/modules/hotel/hotel.controller.ts [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        return reply.badRequest(await error_message(e));
+    }
+}
+
+export async function browseHotelHandler(
+    request: FastifyRequest,
+    reply: FastifyReply
+) {
+    try {
+        const hotels = await request.redis.rememberJSON<Hotel[]>("memCachedAllHotels", 1800, async () => {
```
[the remark this answers is not in the document]
Comment (Anders164a) [reply src/modules/hotel/hotel.controller.ts:38]: Changes added

[the remark this answers is not in the document]
Comment (Anders164a) [reply src/modules/hotel/hotel.controller.ts:38]: Changes added

[the remark this answers is not in the document]
Comment (Anders164a) [reply src/modules/hotel/hotel.controller.ts:38]: Changes added

File: src/modules/hotel/hotel.controller.ts [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+        reply.code(204).send(hotel);
+    } catch (e) {
+        return reply.badRequest(await error_message(e));
+    }
+}
+
+export async function removeRedisCache(request: FastifyRequest, key: string) {
+    await request.redis.del(key);
+}
```
[the remark this answers is not in the document]
Comment (Anders164a) [reply src/modules/hotel/hotel.controller.ts:101]: Added new call in redis.ts that invalidatesCaches

File: src/plugins/redis.ts [hunk 8]
```diff
@@ -50,6 +53,12 @@ export default fastifyPlugin(
             }));
         };
 
+        fastify.redis.invalidateCaches = async (...keys) => {
+            keys.forEach(async (key) => {
+                await fastify.redis.del(key);
+            });
+        }
```
Comment (kristianbinau) [inline src/plugins/redis.ts:60]: @frederikpyt

File: src/plugins/redis.ts [hunk 9]
```diff
@@ -1,4 +1,4 @@
-import { FastifyInstance, FastifyPluginOptions } from "fastify";
+import { FastifyInstance, FastifyPluginOptions, FastifyRequest } from "fastify";
```
Comment (kristianbinau) [inline src/plugins/redis.ts:1]: ARE YOU EVEN USING THIS????????????

File: src/modules/hotel/hotel.controller.ts [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+    DeleteHotelParams
+} from "./hotel.schema";
+import { error_message } from "./hotel.errors";
+import { Hotel } from "@prisma/client";
+
+// In Seconds
+const cache_ttl = 1800;
+
+const cache_key_hotels = "allHotels";
+const cache_key_hotel = "hotel";
```
Comment (kristianbinau) [inline src/modules/hotel/hotel.controller.ts:22]: All consts like this should be all caps.

```suggestion
const CACHE_TTL = 1800;

const CACHE_KEY_HOTELS = "allHotels";
const CACHE_KEY_HOTEL = "hotel";
```

**Review by kristianbinau — CHANGES_REQUESTED** [review]
Just a few

File: src/plugins/redis.ts (hunk 8 above)
Comment (kristianbinau) [reply src/plugins/redis.ts:60]: He will help

File: src/modules/hotel/hotel.controller.ts (hunk 10 above)
Comment (Anders164a) [reply src/modules/hotel/hotel.controller.ts:22]: Changes added

**Review by kristianbinau — APPROVED** [review]
Awesome! 🪨 Turned out great in the end. 🥇

**Pull request merged**